src/aspire/abinitio/commonline_cn.py

```diff
@@ -0,0 +1,367 @@
+        )
+
+        self.n_points_sphere = n_points_sphere
+
+    def _check_symmetry(self, symmetry):
+        if symmetry is None:
+            raise NotImplementedError(
+                "Symmetry type not supplied. Please indicate symmetry."
+            )
+        else:
+            symmetry = symmetry.upper()
+            if not symmetry[0] == "C":
+                raise NotImplementedError(
+                    f"Only Cn symmetry supported. {symmetry} was supplied."
+                )
+            self.order = int(symmetry[1:])
+
+    def _estimate_relative_viewing_directions(self):
+        n_img = self.n_img
+        n_theta = self.n_theta
+        pf = self.pf
+        max_shift_1d = self.max_shift
+        shift_step = self.shift_step
+
+        # Generate candidate rotation matrices and the common-line and
+        # self-common-line indices induced by those rotations.
+        Ris_tilde, R_theta_ijs = self._generate_cand_rots()
+        cijs_inds = self._compute_cls_inds(Ris_tilde, R_theta_ijs)
+        scls_inds = self._compute_scls_inds(Ris_tilde)
+        n_cands = len(Ris_tilde)
+        n_theta_ijs = len(R_theta_ijs)
+
+        # Generate shift phases.
+        r_max = pf.shape[-1]
+        shifts, shift_phases, _ = self._generate_shift_phase_and_filter(
+            r_max, max_shift_1d, shift_step
+        )
+        n_shifts = len(shifts)
+        all_shift_phases = shift_phases.T
+
+        # Reconstruct full polar Fourier for use in correlation.
+        pf /= norm(pf, axis=2)[..., np.newaxis]  # Normalize each ray.
+        pf_full = np.concatenate((pf, np.conj(pf)), axis=1)
+
+        # Step 1: pre-calculate the likelihood with respect to the self-common-lines.
+        scores_self_corrs = np.zeros((n_img, n_cands), dtype=self.dtype)
+        logger.info("Computing likelihood wrt self common-lines.")
+        for i in trange(n_img):
+            pf_i = pf[i]
+            pf_full_i = pf_full[i]
+
+            # Generate shifted versions of image.
+            pf_i_shifted = np.array(
+                [pf_i * shift_phase for shift_phase in all_shift_phases]
+            )
+            pf_i_shifted = np.reshape(pf_i_shifted, (n_shifts * n_theta // 2, r_max))
+
+            # Ignore dc-component.
+            pf_full_i[:, 0] = 0
+            pf_i_shifted[:, 0] = 0
+
+            # Compute correlation of pf_i with itself over all shifts.
+            corrs = pf_i_shifted @ np.conj(pf_full_i).T
+            corrs = np.reshape(corrs, (n_shifts, n_theta // 2, n_theta))
+            corrs_cands = np.max(
+                np.real(corrs[:, scls_inds[:, :, 0], scls_inds[:, :, 1]]), axis=0
+            )
+
+            scores_self_corrs[i] = np.mean(np.real(corrs_cands), axis=1)
+
+        # Remove candidates that are equator images. Equator candidates induce collinear
+        # self common-lines, which always have perfect correlation.
+        # TODO: Should the threshold be parameter-dependent instead of set to 10 degrees?
```

Comment:
Yes please, add as parameter to __init__ (and a docstring), keep their default.

Reply:
Added.